src/main/java/org/elasticsearch/indices/store/IndicesStore.java

```diff
@@ -333,8 +337,63 @@ public String executor() {
-            channel.sendResponse(new ShardActiveResponse(shardActive(request), clusterService.localNode()));
+        public void messageReceived(final ShardActiveRequest request, final TransportChannel channel) throws Exception {
+            // make sure shard is really there before register cluster state obsever
+            ShardId shardId = request.shardId;
+            IndexService indexService = indicesService.indexService(shardId.index().getName());
+            if (indexService != null && indexService.indexUUID().equals(request.indexUUID)) {
+                if (indexService.hasShard(shardId.id()) == false) {
+                    channel.sendResponse(new ShardActiveResponse(false, clusterService.localNode()));
+                }
+            } else {
+                channel.sendResponse(new ShardActiveResponse(false, clusterService.localNode()));
+            }
+            // create observer just in case
+            ClusterStateObserver observer = new ClusterStateObserver(clusterService, request.timeout, logger);
+            // check if shard is active. if so, all is good
+            boolean shardActive = shardActive(request);
+            if (shardActive) {
+                channel.sendResponse(new ShardActiveResponse(shardActive(request), clusterService.localNode()));
+            } else {
+                // shard is not active, might be POST_RECOVERY so check if cluster state changed inbetween or wait for next change
+                observer.waitForNextChange(new ClusterStateObserver.Listener() {
+                    @Override
+                    public void onNewClusterState(ClusterState state) {
+                        sendResultOrLogError(shardActive(request));
+                    }
+
+                    @Override
+                    public void onClusterServiceClose() {
+                        sendResultOrLogError(false);
+                    }
+
+                    @Override
+                    public void onTimeout(TimeValue timeout) {
+                        sendResultOrLogError(shardActive(request));
+                    }
+
+                    public void sendResultOrLogError(boolean shardActive) {
+                        try {
+                            channel.sendResponse(new ShardActiveResponse(shardActive, clusterService.localNode()));
+                        } catch (IOException e) {
+                            logger.error("failed send response for shard active while trying to delete shard {} - shard will probably not be removed", e, request.shardId);
+                        } catch (EsRejectedExecutionException e) {
+                            logger.error("failed send response for shard active while trying to delete shard {} - shard will probably not be removed", e, request.shardId);
+                        }
+                    }
+                }, new ClusterStateObserver.ChangePredicate() {
```

Comment:
can we inherit ValidationPredicate ? it takes care of all the subtleties and requires implementing one method